File: solutions/lua/binary-search/2/binary-search.lua
return function(array, target)
  local function binary_search(low, high)
    local avg = (low + high) // 2
    return low > high and -1 or
        target < array[avg] and binary_search(low, avg - 1) or
        target > array[avg] and binary_search(avg + 1, high) or avg
  end
  return binary_search(1, #array)
end

File: solutions/ruby/binary-search/2/binary_search.rb
class BinarySearch
  def initialize(array)= @array = array

  def search_for(target, low = 0, high = @array.size - 1)
    avg = (low + high) / 2
    if low > high then nil
    elsif target < @array[avg] then search_for(target, low, avg - 1)
    elsif target > @array[avg] then search_for(target, avg + 1, high)
    else avg
    end
  end
end
